File: practice/algorithms/implementation/the_time_in_words/the_time_in_words.py
#!/bin/python

import sys

def to_word(i_h):
    w = ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten',
         'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 
         'eighteen', 'nineteen', 'twenty']
    return w[i_h - 1]

h = int(raw_input().strip())
m = int(raw_input().strip())

if m == 0:
    print to_word(h) + ' o\' clock'
elif m == 30:
        print 'half past ' + to_word(h)
elif m < 30:
    if m == 1:
        print 'one minute past ' + to_word(h)
    elif m == 15:
        print 'quarter past '  + to_word(h)
    elif m < 21:
        print to_word(m) + ' minutes past ' + to_word(h)
    elif m < 30:
        print 'twenty ' + to_word(m % 10) + ' minutes past ' + to_word(h)
else:
    m = 60 - m
    if m == 1:
        print 'one to ' + to_word(h + 1)
    elif m == 15:
        print 'quarter to '  + to_word(h + 1)
    elif m < 21:
        print to_word(m) + ' minutes to ' + to_word(h + 1)
    elif m < 30:
        print 'twenty ' + to_word(m % 10) + ' minutes to ' + to_word(h + 1)

File: practice/algorithms/implementation/the_time_in_words/the_time_in_words.hs
import Control.Applicative
import Control.Monad
import System.IO

toWord value
    | value == 1  = "one"
    | value == 2  = "two"
    | value == 3  = "three"
    | value == 4  = "four"
    | value == 5  = "five"
    | value == 6  = "six"
    | value == 7  = "seven"
    | value == 8  = "eight"
    | value == 9  = "nine"
    | value == 10 = "ten"
    | value == 11 = "eleven"
    | value == 12 = "twelve"
    | value == 13 = "thirteen"
    | value == 14 = "fourteen"
    | value == 15 = "fifteen"
    | value == 16 = "sixteen"
    | value == 17 = "seventeen"
    | value == 18 = "eighteen"
    | value == 19 = "nineteen"
    | value == 20 = "twenty"
    
timeInWords h m
    | m == 0  =  toWord h ++ " o' clock"
    | m == 1  = "one minute past " ++ toWord h
    | m == 15 = "quarter past " ++ toWord h
    | m < 21  = toWord m ++ " minutes past " ++ toWord h
    | m < 30  = "twenty " ++ (toWord $ m `mod` 10) ++ " minutes past " ++ toWord h
    | m == 30 = "half past " ++ toWord h
    | m == 59 = "one to " ++ toWord (h + 1)
    | m == 45 = "quarter to " ++ toWord (h + 1)
    | m > 39  = toWord (60 - m) ++ " minutes to " ++ toWord (h + 1)
    | m > 30  = "twenty " ++ (toWord $ (60 - m) `mod` 10) ++ " minutes to " ++ toWord (h + 1)

main :: IO ()
main = do
    h <- readLn :: IO Int
    m <- readLn :: IO Int
    putStr $ timeInWords h m

